#4524 Fix active extension availability check

BLoe opened the pull request:
## What does this PR do?

- Fixes #4524 

## Demo

https://www.loom.com/share/1760c26071594d28ad6da3728184cd27

## Future Work

- There is some investigation required around possible issues with availability calculations on SPA websites
- We might want to tweak how the items appear/disappear when the checkbox is toggled, but some UX thought is needed there

## Checklist

- [ ] Add tests - created a separate refactoring ticket for sidebar UI tests: #4626
- [ ] Run Storybook and manually confirm that all stories are working
- [x] Designate a primary reviewer - @BALEHOK

BALEHOK commented:
Hey @BLoe 
I see a weird behavior of the hidden items. If select a hidden extension, then switch off the checkbox, and then select another extension, the previously selected one does not disappear. Is this a bug or a feature?
https://www.loom.com/share/d735259787a84c3ea316d95e13420983

Does this PR only fixes the unavailable extensions count increment?

BALEHOK commented:
How about adding tests for the counter, checkbox, and visible items?
We've got multiple issues with this part of the Page Editor. I believe we need to cover it with unit and/or UI tests to shield ourselves from same or similar bugs in the future.

fregante commented on src/pageEditor/slices/editorSlice.ts:
```diff
@@ -226,14 +225,45 @@ const checkAvailableDynamicElements = createAsyncThunk<
+    .filter(
+      (extension) => !elements.some((element) => element.uuid === extension.id)
+    )
+    .map((extension) => extension.id);
```
We're only using the length of this array on the end, I think this line can be avoided altogether.

BLoe replied:
Would code like this be more clear?
```
  // Find installed extensions that have not been selected yet, i.e. do
  // not have matching dynamic elements
  const installedNotDynamicIds = differenceBy(installedExtensions, dynamicElements, (extensionOrElement) => {
    if (isExtension(extensionOrElement)) {
      return extensionOrElement.id;
    }

    return extensionOrElement.uuid;
  });
```

I'm wondering if it's the code that's confusing you or just the general awkwardness/weirdness of the state design and logic here.

fregante replied:
This helps a lot 👍

BLoe replied:
Ok, cool! Pushing a commit shortly...

fregante commented on src/pageEditor/slices/editorSlice.ts:
```diff
@@ -226,14 +225,45 @@ const checkAvailableDynamicElements = createAsyncThunk<
+    .map((extension) => extension.id);
+  const availableInstalledNotDynamicIds = availableInstalledIds.filter((id) =>
+    installedNotDynamicIds.includes(id)
+  );
```
This looks like an intersection, lodash has a helper for this. Generally that makes the final intent more obvious and the code shorter

fregante reviewed (commented):
I'm not familiar with the requirements here, but I feel that this code can be shortened, the logic is a bit difficult to follow (and can thus hide further bugs)

fregante commented on src/pageEditor/slices/editorSlice.ts:
```diff
@@ -226,14 +225,45 @@ const checkAvailableDynamicElements = createAsyncThunk<
+    ? uniq([activeElement.uuid, ...state.editor.availableDynamicIds])
+    : state.editor.availableDynamicIds.filter(
+        (id) => id !== activeElement.uuid
+      );
```
I think this filter can be replaced with `_.pull(state.editor.availableDynamicIds, activeElement.uuid)` or some similar filtering function

BLoe commented:
> If select a hidden extension, then switch off the checkbox, and then select another extension, the previously selected one does not disappear. Is this a bug or a feature?

If a recipe is open then it won't disappear when the checkbox is unselected. If you try that with a single extension it will disappear when you select the other one.

BLoe commented:
@fregante @BALEHOK I refactored the code a bit and added a bunch of comments. We already have some testing around the `arrangeElements` function. I started looking into setting up UI tests for the sidebar but I need to stop working on that for right now and do some other things this week, so I created a follow-up refactoring ticket for sidebar UI tests (linked in the PR description). I tested this manually and didn't see any issues. The way extensions and recipes appear/disappear might seem confusing, but as far as I can tell it's all how it's expected to work.

BALEHOK commented on src/pageEditor/slices/editorSlice.ts:
```diff
@@ -782,23 +851,21 @@ export const editorSlice = createSlice({
-          state.unavailableDynamicCount =
-            elements.length - state.availableDynamicIds.length;
-        }
+        ) => ({
+          ...state,
+          availableDynamicIds,
+          unavailableDynamicCount,
+          unavailableInstalledCount,
+        })
```
recreating a state object is not required. simple re-assignment of these 3 properties should work just as well

BALEHOK commented on src/pageEditor/slices/editorSlice.ts:
```diff
@@ -227,14 +236,74 @@ const checkAvailableDynamicElements = createAsyncThunk<
 });
 
 const checkActiveElementAvailability = createAsyncThunk<
```
I've seen your comment regarding a dedicated ticket for refactoring of UI tests.

However this thunk has a lot of logic. It can be **unit** tested. Separately, without UI or whatsoever. You can check this one for reference https://github.com/pixiebrix/pixiebrix-extension/pull/4596/files#diff-4f1725cd753df85d0a5edf73af716a5f1296659d5db865023cda8508239ebff7

BLoe replied:
I added a unit test 👍

BALEHOK reviewed (approved):
Let's move on with this
